bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

```diff
@@ -77,19 +85,49 @@ public ChannelPool(List<HeaderInterceptor> headerInterceptors, ChannelFactory fa
    */
   public void ensureChannelCount(int capacity) throws IOException {
     if (totalSize.get() < capacity) {
-      synchronized(this){
+      synchronized (totalSize) {
         if (totalSize.get() < capacity) {
-          List<Channel> newChannelList = new ArrayList<>(channels.get());
-          while(newChannelList.size() < capacity) {
-            newChannelList.add(factory.create());
-          }
-          setChannels(newChannelList);
+          channels.set(
+              ImmutableList.<Channel> builder()
+                  .addAll(channels.get())
+                  .addAll(createNewChannels(capacity - totalSize.get()))
+                  .build()
+          );
           totalSize.set(capacity);
         }
       }
     }
   }
 
+  /**
+   * Creates a new list of {@link Channel} to be used in parallel operations.
+   * @param newChannelCount The number of new Channels to create.
+   * @return a new List of {@link Channel}
+   * @throws IOException if an exception occurred during Channel creation.
+   */
+  protected List<Channel> createNewChannels(int newChannelCount) throws IOException {
+    ExecutorService executorService = BigtableSessionSharedThreadPools.getInstance().getBatchThreadPool();
+    ListeningExecutorService listeningExecutor = MoreExecutors.listeningDecorator(executorService);
+
+    Callable<Channel> createChannelCallable = new Callable<Channel>() {
+      @Override
+      public Channel call() throws Exception {
+        return factory.create();
```

Comment:
Could you add a comment with how long does it take to create a channel? In general all these optimizations should have a sense of how much time they save at the time they were implemented. They might become obsolete at some point and we will not know if we can clean them up.